rust/connlib/libs/common/src/session.rs

```
@@ -46,34 +50,123 @@ pub trait ControlSession<T, CB: Callbacks> {
 /// A session is created using [Session::connect], then to stop a session we use [Session::disconnect].
 pub struct Session<T, U, V, R, M, CB: Callbacks> {
     runtime: Arc<Mutex<Option<Runtime>>>,
-    callbacks: CB,
+    callbacks: CallbackErrorFacade<CB>,
     _phantom: PhantomData<(T, U, V, R, M)>,
 }
 
 /// Traits that will be used by connlib to callback the client upper layers.
 pub trait Callbacks: Clone + Send + Sync {
+    /// Error returned when a callback fails.
+    type Error: Debug + Display + StdError;
+
     /// Called when the tunnel address is set.
     fn on_set_interface_config(
         &self,
         tunnel_address_v4: Ipv4Addr,
         tunnel_address_v6: Ipv6Addr,
         dns_address: Ipv4Addr,
-    );
+    ) -> StdResult<(), Self::Error>;
     /// Called when the tunnel is connected.
-    fn on_tunnel_ready(&self);
+    fn on_tunnel_ready(&self) -> StdResult<(), Self::Error>;
     /// Called when when a route is added.
-    fn on_add_route(&self, route: String);
+    fn on_add_route(&self, route: IpNetwork) -> StdResult<(), Self::Error>;
     /// Called when when a route is removed.
-    fn on_remove_route(&self, route: String);
+    fn on_remove_route(&self, route: IpNetwork) -> StdResult<(), Self::Error>;
     /// Called when the resource list changes.
-    fn on_update_resources(&self, resource_list: Vec<ResourceDescription>);
+    fn on_update_resources(
+        &self,
+        resource_list: Vec<ResourceDescription>,
+    ) -> StdResult<(), Self::Error>;
     /// Called when the tunnel is disconnected.
     ///
     /// If the tunnel disconnected due to a fatal error, `error` is the error
     /// that caused the disconnect.
-    fn on_disconnect(&self, error: Option<&Error>);
+    fn on_disconnect(&self, error: Option<&Error>) -> StdResult<(), Self::Error>;
```

I think we always ignore this error, shouldn't we just get rid of it?